python/cuBERT/pybind/cuBERT_pybind.cc

```diff
@@ -0,0 +1,314 @@
+      auto float_free = py::capsule(output, [](void *f) {
+        float *output = static_cast<float *>(f);
+        delete[] output;
+      });
+
+      return py::array_t<float>({batch_size}, {sizeof(float)}, (float *)output, float_free);  // numpy array
+    },
+    py::arg("model"),
+    py::arg("batch_size"),
+    py::arg("input_ids"),
+    py::arg("input_mask"),
+    py::arg("segment_ids"),
+    py::arg("output_type") = cuBERT_OutputType::cuBERT_LOGITS
+  );
+
+  m.def(
+    "cuBERT_compute_half_float",
+    [](
+      void *model,
+      int batch_size,
+      std::vector<int> input_ids,
+      std::vector<int> input_mask,
+      std::vector<int> segment_ids,
+      cuBERT_OutputType output_type
+    ) {
+      void *output = (void *)(new half_float::half[batch_size]);
+
+      char *input_mask_buf = int2char(input_mask);
+      char *segment_ids_buf = int2char(segment_ids);
+
+      cuBERT_compute(
+        model,
+        batch_size,
+        input_ids.data(),
+        input_mask_buf,
+        segment_ids_buf,
+        output,
+        output_type,
+        cuBERT_ComputeType::cuBERT_COMPUTE_HALF
+      );
+
+      delete_ptr(input_mask_buf);
+      delete_ptr(segment_ids_buf);
+
+      // Create a Python object that will free the allocated memory when destroyed:
+      auto half_float_free = py::capsule(output, [](void *f) {
+        half_float::half *output = static_cast<half_float::half *>(f);
+        delete[] output;
+      });
+
+      return py::array_t<half_float::half>(
+        {batch_size}, {sizeof(half_float::half)}, (half_float::half *)output, half_float_free
+      );
+    },
+    py::arg("model"),
+    py::arg("batch_size"),
+    py::arg("input_ids"),
+    py::arg("input_mask"),
+    py::arg("segment_ids"),
+    py::arg("output_type") = cuBERT_OutputType::cuBERT_LOGITS
+  );
+
+  m.def(
+    "cuBERT_close",
+    [](
+      void *model,
+      cuBERT_ComputeType compute_type
+    ) {
+      cuBERT_close(model, compute_type);
+    },
+    py::arg("model"),
+    py::arg("compute_type") = cuBERT_ComputeType::cuBERT_COMPUTE_FLOAT
+  );
+
+  /** high level API including tokenization **/
+  m.def(
+    "cuBERT_open_tokenizer",
+    [](
+      const char *vocab_file,
+      int do_lower_case = 1
+    ) {
+      return cuBERT_open_tokenizer(vocab_file, do_lower_case);
+    },
+    py::arg("vocab_file"),
+    py::arg("do_lower_case") = 1
+  );
+
+  m.def(
+    "cuBERT_close_tokenizer",
+    [](void *tokenizer) { cuBERT_close_tokenizer(tokenizer); },
+    py::arg("tokenizer")
+  );
+
+  m.def(
+    "cuBERT_tokenize_compute_float",
+    [](
+      void *model,
+      void *tokenizer,
+      int batch_size,
+      std::vector<std::string> text_a,
+      std::vector<std::string> text_b,
+      cuBERT_OutputType output_type
+    ) {
+      char **arr_text_a = str2char(text_a);
+      char **arr_text_b = str2char(text_b);
+
+      void *output = (void *)(new float[batch_size]);
```

Comment:
I think the `output` could be allocated in python side, and pass the `void*` pointer into c++. There are several benefits:

1. `cuBERT_tokenize_compute_float` and `cuBERT_tokenize_compute_half_float` could be combined together. Because the major difference is the calling of `return py::array_t<float>(...)`, which could be avoided.

2. the array size `new float[batch_size]` is depending on output_type. If output_type is pooled_output, the size `batch_size` here is wrong. So it is better for python code to pre-allocate enough memory for our c++, and we just fill the output buffer.